packages/rest/src/body-parser.ts

```diff
@@ -0,0 +1,287 @@
+
+    const textOptions = Object.assign(
+      {type: 'text/*', limit: DEFAULT_LIMIT},
+      getParserOptions('text', options),
+    );
+    this.textParser = text(textOptions);
+
+    this.parsers = parsers.concat([
+      {
+        name: 'json',
+        supports: mediaType => !!is(mediaType, 'json'),
+        parse: request => this.parseJsonBody(request),
+      },
+      {
+        name: 'urlencoded',
+        supports: mediaType => !!is(mediaType, 'urlencoded'),
+        parse: request => this.parseUrlencodedBody(request),
+      },
+      {
+        name: 'text',
+        supports: mediaType => !!is(mediaType, 'text/*'),
+        parse: request => this.parseTextBody(request),
+      },
+    ]);
+  }
+
+  async parseJsonBody(request: Request) {
+    await parse(this.jsonParser, request);
+    return {value: request.body};
+  }
+
+  async parseUrlencodedBody(request: Request) {
+    await parse(this.urlencodedParser, request);
+    return {value: request.body, coercionRequired: true};
+  }
+
+  async parseTextBody(request: Request) {
+    await parse(this.textParser, request);
+    return {value: request.body};
+  }
+
+  private normalizeParsingError(err: HttpError) {
+    debug('Cannot parse request body %j', err);
+    if (!err.statusCode || err.statusCode >= 500) {
+      err.statusCode = 400;
+    }
+    return err;
+  }
+
+  async loadRequestBodyIfNeeded(
+    operationSpec: OperationObject,
+    request: Request,
+    options: RequestBodyParserOptions = {},
+  ): Promise<RequestBody> {
+    const requestBody: RequestBody = {
+      value: undefined,
+    };
+    if (!operationSpec.requestBody) return requestBody;
+
+    debug('Request body parser options: %j', options);
+
+    const contentType = getContentType(request) || 'application/json';
+    debug('Loading request body with content type %j', contentType);
+
+    // the type of `operationSpec.requestBody` could be `RequestBodyObject`
+    // or `ReferenceObject`, resolving a `$ref` value is not supported yet.
+    if (isReferenceObject(operationSpec.requestBody)) {
+      throw new Error('$ref requestBody is not supported yet.');
+    }
+
+    let content = operationSpec.requestBody.content || {};
+    if (!Object.keys(content).length) {
+      content = {
+        // default to allow json and urlencoded
+        'application/json': {schema: {type: 'object'}},
+        'application/x-www-form-urlencoded': {schema: {type: 'object'}},
+      };
+    }
+
+    // Check of the request content type matches one of the expected media
+    // types in the request body spec
+    let matchedMediaType: string | false = false;
+    for (const type in content) {
+      matchedMediaType = is(contentType, type);
+      if (matchedMediaType) {
+        debug('Matched media type: %s -> %s', type, contentType);
+        requestBody.mediaType = type;
+        requestBody.schema = content[type].schema;
+        // Skip body parsing as the controller method wants to have full control
```

Comment:
Now that we allow custom body parsers, I am no longer convinced it's a good idea to allow controllers to have full control over body parsing.

What uses case do you have for this feature? In what circumstances would an application or an extension need to parse the body directly and would not be able to install custom body parser instead?

Reply:
I don't think the parsers can handle all cases or are optimal. Allowing a controller method to fully control how it's handled is desirable, for example, a method may want to streamline large json request processing.

Reply:
I was thinking about this some more and I agree with you that this feature makes sense. I'd like us to reframe it a bit though. Instead of focusing on implementation details (skipping body parsing) & describing what the app/extension developer does not want to, can we describe this feature from the point of what they actually want to achieve? I.e. receive the request body as a stream.

In that light, I am proposing to:

Find a different way how developers can tell the framework that they want to receive the request body as a (raw/original) stream. 
- At minimum, we should rename `x-skip-body-parsing`, e.g. to something like `x-stream: true`. I think it will be even better to use an enum instead of a boolean flag, e.g. `x-process-as: stream`. It will allow us to provide other ways for consuming request bodies, e.g. consume large JSON arrays via async iterators (`x-process-as: iterator`). I cannot come up with a good name for this new option right now, please try to find a better one yourself.
- Consider adding a higher-level sugar API at `@requestBody` decorator. I am not sure how easy this would be to design and I am fine to leave it out of scope of this pull request, but we should at least consider the user/developer experience of people using this feature in their projects.

In the documentation for this new feature (which you still need to add AFAICT), use the scenario where the app/extension developer wants to consume the request body as a stream to drive the narrative. I can imagine this new documentation can take a form of a new page in the section "How tos", plus a short mention of the new extension flag in apidocs and/or `@requestBody` overview.